[SPARK-32804][Launcher] Fix run-example command builder bug

KevinSmile opened the pull request:
<!--
Thanks for sending a pull request!  Here are some tips for you:
  1. If this is your first time, please read our contributor guidelines: https://spark.apache.org/contributing.html
  2. Ensure you have added or run the appropriate tests for your PR: https://spark.apache.org/developer-tools.html
  3. If the PR is unfinished, add '[WIP]' in your PR title, e.g., '[WIP][SPARK-XXXX] Your PR title ...'.
  4. Be sure to keep the PR description updated to reflect all changes.
  5. Please write your PR title to summarize what this PR proposes.
  6. If possible, provide a concise example to reproduce the issue for a faster review.
  7. If you want to add a new configuration, please read the guideline first for naming configurations in
     'core/src/main/scala/org/apache/spark/internal/config/ConfigEntry.scala'.
-->

### What changes were proposed in this pull request?
<!--
Please clarify what changes you are proposing. The purpose of this section is to outline the changes and how this PR fixes the issue. 
If possible, please consider writing useful notes for better and faster reviews in your PR. See the examples below.
  1. If you refactor some codes with changing classes, showing the class hierarchy will help reviewers.
  2. If you fix some SQL features, you can provide some references of other DBMSes.
  3. If there is design documentation, please add the link.
  4. If there is a discussion in the mailing list, please add the link.
-->
Bug fix in run-example command builder (as described in [SPARK-32804], run-example failed in standalone-cluster mode): 
1. Missing primaryResource arg.
2. Wrong appResource arg.

which will affect `SparkSubmit` in Standalone-Cluster mode: 
https://github.com/apache/spark/blob/32d87c2b595b4aac2d9274424a43697299638f61/core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala#L695-L696

and get error at:
https://github.com/apache/spark/blob/f55694638d45f34ab91f6f6ec2066cbf7631f4af/core/src/main/scala/org/apache/spark/deploy/ClientArguments.scala#L74-L89

### Why are the changes needed?
<!--
Please clarify why the changes are needed. For instance,
  1. If you propose a new API, clarify the use case for a new API.
  2. If you fix a bug, you can clarify why it is a bug.
-->
Bug: run-example failed in standalone-cluster mode

### Does this PR introduce _any_ user-facing change?
<!--
Note that it means *any* user-facing change including all aspects such as the documentation fix.
If yes, please clarify the previous behavior and the change this PR proposes - provide the console output, description and/or an example to show the behavior difference if possible.
If possible, please also clarify if this is a user-facing change compared to the released Spark versions or within the unreleased branches such as master.
If no, write 'No'.
-->
Yes. User can run-example in standalone-cluster mode now.

### How was this patch tested?
<!--
If tests were added, say they were added here. Please make sure to add some test cases that check the changes thoroughly including negative and positive cases if possible.
If it was tested in a way different from regular unit tests, please clarify how you tested step by step, ideally copy and paste-able, so that other reviewers can test and check, and descendants can verify in the future.
If tests were not added, please describe why they were not added and/or why it was difficult to add.
-->
New ut added.
Also it's a user-facing bug, so better re-check the real case in [SPARK-32804].

KevinSmile committed:
[SPARK-32804][Launcher] fix run-example command builder bug

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

srowen commented:
Are you sure? this already adds the example JAR. You have to specify the name of the example to run. Why would you give the JAR as an arg?

KevinSmile commented:
@srowen 

Currently, when you do `run-example`,  `SparkSubmitCommandBuilder` will auto add `--jars` (set to all jars in `examples/jars` folder) for you. 
https://github.com/apache/spark/blob/f5360e761ef161f7e04526b59a4baf53f1cf8cd5/launcher/src/main/java/org/apache/spark/launcher/SparkSubmitCommandBuilder.java#L215-L217

However, `--jars` is different from `<application-jar>` (aka primaryResource, which contains your Main-Class) .

So when it (in standalone cluster mode) goes to:
https://github.com/apache/spark/blob/32d87c2b595b4aac2d9274424a43697299638f61/core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala#L695-L696

https://github.com/apache/spark/blob/f55694638d45f34ab91f6f6ec2066cbf7631f4af/core/src/main/scala/org/apache/spark/deploy/ClientArguments.scala#L74-L89

We get the error described in [SPARK-32804]:
1. Missing primaryResource arg. ====> aka `<application-jar>`
2. Wrong appResource arg. ====> "spark-internal" should not be set as an arg in `run-example`
Actually, in [SPARK-32804], "spark-internal" is treated as primaryResource
![image](https://user-images.githubusercontent.com/17903517/92329159-2bdfbd80-f098-11ea-8911-b2dfc7421dcd.png)

(But client mode does not have the problem.)

I would say it's kind of tricky......

KevinSmile commented:
Or we can check this doc, as `run-example` will use `spark-submit`:
![image](https://user-images.githubusercontent.com/17903517/92329276-1b7c1280-f099-11ea-9534-8f1a1b21be25.png)

srowen commented:
You shouldn't have to pass the app jar if you are specifying the class name. 
I can't reproduce it -- except on standalone + cluster mode. I actually am surprised that is a valid combination.
I take your point here, that there does not seem to be a good reason to put the dummy 'spark-internal' value on the command line. How about just omitting it if it has this special value?

KevinSmile commented on launcher/src/main/java/org/apache/spark/launcher/SparkSubmitCommandBuilder.java:
```diff
@@ -241,9 +241,11 @@
     args.addAll(parsedArgs);
+
     if (appResource != null) {
       args.add(appResource);
```
When you do `run-example`,  you just specify the class name(e.g. SparkPi),  and no need to specify the app-jar. So in backend code, `appResource` should be auto-find-and-set as the examples' main app jar (e.g. `./examples/jars/spark-examples_2.12-3.0.0.jar`), and then be added as an arg ( this arg will later be set as the so-called `primaryResource`).

KevinSmile commented:
When you do `run-example`,  you just specify the class name(e.g. SparkPi),  and no need to specify the app-jar. So in backend code, `appResource` should be auto-find-and-set as the examples' main app jar (e.g. `./examples/jars/spark-examples_2.12-3.0.0.jar`), and then be added as an arg ( this arg will later be set as the so-called `primaryResource`).

srowen commented:
Yes, I'm talking about how `spark-submit` works, not `run-example`. In this case it should specify `--class`, as the JAR is already present. Specifying a JAR is for when you want it to read Main-Class from the Manifest. It's worth seeing if that works, just not specifying it, as it should already set `--class` in this case.

KevinSmile commented:
`app-jar` arg is always needed in `spark-submit` user command line, backend code won't add it for you; 
while for `run-example`,  the backend code should auto-add `app-jar` arg for you.

`--jars` option and `app-jar` arg are two different things.

![image](https://user-images.githubusercontent.com/17903517/92422250-2759f880-f1af-11ea-8cd0-34f8e45046d2.png)

srowen commented:
I think you're probably right about this, but then how does this work at all? the argument is bogus and ignored in most deployment scenarios? that would make sense.

KevinSmile commented:
I think it only affects run-example in standalone-cluster mode.

As client mode use a different logic.

And for `spark-submit`,  if `--jars` is provided(including your main jar),  you will still get the right result, but with a log error.


```
./bin/spark-submit \
>   --class org.apache.spark.examples.SparkPi \
>   --jars ./examples/jars/spark-examples_2.12-3.0.0.jar,./examples/jars/scopt_2.12-3.7.1.jar \
>   --master spark://KevinMac.local:7077 \
>   --deploy-mode client \
>   100
20/09/08 08:52:28 WARN NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable
Using Spark's default log4j profile: org/apache/spark/log4j-defaults.properties
20/09/08 08:52:28 WARN DependencyUtils: Local jar /Users/Kevin/Development/learnSpark/spark-3.0.0-bin-hadoop2.7/100 does not exist, skipping.
20/09/08 08:52:29 INFO SparkContext: Running Spark version 3.0.0
20/09/08 08:52:29 INFO ResourceUtils: ==============================================================
20/09/08 08:52:29 INFO ResourceUtils: Resources for spark.driver:

20/09/08 08:52:29 INFO ResourceUtils: ==============================================================
20/09/08 08:52:29 INFO SparkContext: Submitted application: Spark Pi
20/09/08 08:52:29 INFO SecurityManager: Changing view acls to: Kevin
20/09/08 08:52:29 INFO SecurityManager: Changing modify acls to: Kevin
20/09/08 08:52:29 INFO SecurityManager: Changing view acls groups to:
20/09/08 08:52:29 INFO SecurityManager: Changing modify acls groups to:
20/09/08 08:52:29 INFO SecurityManager: SecurityManager: authentication disabled; ui acls disabled; users  with view permissions: Set(Kevin); groups with view permissions: Set(); users  with modify permissions: Set(Kevin); groups with modify permissions: Set()
20/09/08 08:52:29 INFO Utils: Successfully started service 'sparkDriver' on port 62182.
20/09/08 08:52:29 INFO SparkEnv: Registering MapOutputTracker
20/09/08 08:52:29 INFO SparkEnv: Registering BlockManagerMaster
20/09/08 08:52:29 INFO BlockManagerMasterEndpoint: Using org.apache.spark.storage.DefaultTopologyMapper for getting topology information
20/09/08 08:52:29 INFO BlockManagerMasterEndpoint: BlockManagerMasterEndpoint up
20/09/08 08:52:29 INFO SparkEnv: Registering BlockManagerMasterHeartbeat
20/09/08 08:52:29 INFO DiskBlockManager: Created local directory at /private/var/folders/dg/b9vzh3ls5d57qmmffjpvc8540000gn/T/blockmgr-8ba808a8-d7dc-4e69-901d-578282febf22
20/09/08 08:52:29 INFO MemoryStore: MemoryStore started with capacity 366.3 MiB
20/09/08 08:52:29 INFO SparkEnv: Registering OutputCommitCoordinator
20/09/08 08:52:30 INFO Utils: Successfully started service 'SparkUI' on port 4040.
20/09/08 08:52:30 INFO SparkUI: Bound SparkUI to 0.0.0.0, and started at http://192.168.0.100:4040
20/09/08 08:52:30 INFO SparkContext: Added JAR file:///Users/Kevin/Development/learnSpark/spark-3.0.0-bin-hadoop2.7/examples/jars/spark-examples_2.12-3.0.0.jar at spark://192.168.0.100:62182/jars/spark-examples_2.12-3.0.0.jar with timestamp 1599526350171
20/09/08 08:52:30 INFO SparkContext: Added JAR file:///Users/Kevin/Development/learnSpark/spark-3.0.0-bin-hadoop2.7/examples/jars/scopt_2.12-3.7.1.jar at spark://192.168.0.100:62182/jars/scopt_2.12-3.7.1.jar with timestamp 1599526350172
20/09/08 08:52:30 ERROR SparkContext: Failed to add file:/Users/Kevin/Development/learnSpark/spark-3.0.0-bin-hadoop2.7/100 to Spark environment
java.io.FileNotFoundException: Jar /Users/Kevin/Development/learnSpark/spark-3.0.0-bin-hadoop2.7/100 not found
	at org.apache.spark.SparkContext.addLocalJarFile$1(SparkContext.scala:1827)
	at org.apache.spark.SparkContext.addJar(SparkContext.scala:1881)
	at org.apache.spark.SparkContext.$anonfun$new$11(SparkContext.scala:485)
	at org.apache.spark.SparkContext.$anonfun$new$11$adapted(SparkContext.scala:485)
	at scala.collection.mutable.ResizableArray.foreach(ResizableArray.scala:62)
	at scala.collection.mutable.ResizableArray.foreach$(ResizableArray.scala:55)
	at scala.collection.mutable.ArrayBuffer.foreach(ArrayBuffer.scala:49)
	at org.apache.spark.SparkContext.<init>(SparkContext.scala:485)
	at org.apache.spark.SparkContext$.getOrCreate(SparkContext.scala:2555)
	at org.apache.spark.sql.SparkSession$Builder.$anonfun$getOrCreate$1(SparkSession.scala:930)
	at scala.Option.getOrElse(Option.scala:189)
	at org.apache.spark.sql.SparkSession$Builder.getOrCreate(SparkSession.scala:921)
	at org.apache.spark.examples.SparkPi$.main(SparkPi.scala:30)
	at org.apache.spark.examples.SparkPi.main(SparkPi.scala)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.apache.spark.deploy.JavaMainApplication.start(SparkApplication.scala:52)
	at org.apache.spark.deploy.SparkSubmit.org$apache$spark$deploy$SparkSubmit$$runMain(SparkSubmit.scala:928)
	at org.apache.spark.deploy.SparkSubmit.doRunMain$1(SparkSubmit.scala:180)
	at org.apache.spark.deploy.SparkSubmit.submit(SparkSubmit.scala:203)
	at org.apache.spark.deploy.SparkSubmit.doSubmit(SparkSubmit.scala:90)
	at org.apache.spark.deploy.SparkSubmit$$anon$2.doSubmit(SparkSubmit.scala:1007)
	at org.apache.spark.deploy.SparkSubmit$.main(SparkSubmit.scala:1016)
	at org.apache.spark.deploy.SparkSubmit.main(SparkSubmit.scala)
20/09/08 08:52:30 INFO StandaloneAppClient$ClientEndpoint: Connecting to master spark://KevinMac.local:7077...
20/09/08 08:52:30 INFO TransportClientFactory: Successfully created connection to KevinMac.local/192.168.0.100:7077 after 70 ms (0 ms spent in bootstraps)
20/09/08 08:52:30 INFO StandaloneSchedulerBackend: Connected to Spark cluster with app ID app-20200908085230-0001
20/09/08 08:52:30 INFO StandaloneAppClient$ClientEndpoint: Executor added: app-20200908085230-0001/0 on worker-20200908082422-192.168.0.100-61280 (192.168.0.100:61280) with 4 core(s)
20/09/08 08:52:30 INFO StandaloneSchedulerBackend: Granted executor ID app-20200908085230-0001/0 on hostPort 192.168.0.100:61280 with 4 core(s), 1024.0 MiB RAM
20/09/08 08:52:30 INFO Utils: Successfully started service 'org.apache.spark.network.netty.NettyBlockTransferService' on port 62184.
20/09/08 08:52:30 INFO NettyBlockTransferService: Server created on 192.168.0.100:62184
20/09/08 08:52:30 INFO BlockManager: Using org.apache.spark.storage.RandomBlockReplicationPolicy for block replication policy
20/09/08 08:52:30 INFO StandaloneAppClient$ClientEndpoint: Executor updated: app-20200908085230-0001/0 is now RUNNING
20/09/08 08:52:30 INFO BlockManagerMaster: Registering BlockManager BlockManagerId(driver, 192.168.0.100, 62184, None)
20/09/08 08:52:30 INFO BlockManagerMasterEndpoint: Registering block manager 192.168.0.100:62184 with 366.3 MiB RAM, BlockManagerId(driver, 192.168.0.100, 62184, None)
20/09/08 08:52:30 INFO BlockManagerMaster: Registered BlockManager BlockManagerId(driver, 192.168.0.100, 62184, None)
20/09/08 08:52:30 INFO BlockManager: Initialized BlockManager: BlockManagerId(driver, 192.168.0.100, 62184, None)
20/09/08 08:52:31 INFO StandaloneSchedulerBackend: SchedulerBackend is ready for scheduling beginning after reached minRegisteredResourcesRatio: 0.0
20/09/08 08:52:32 INFO SparkContext: Starting job: reduce at SparkPi.scala:38
20/09/08 08:52:32 INFO DAGScheduler: Got job 0 (reduce at SparkPi.scala:38) with 2 output partitions
20/09/08 08:52:32 INFO DAGScheduler: Final stage: ResultStage 0 (reduce at SparkPi.scala:38)
20/09/08 08:52:32 INFO DAGScheduler: Parents of final stage: List()
20/09/08 08:52:32 INFO DAGScheduler: Missing parents: List()
20/09/08 08:52:32 INFO DAGScheduler: Submitting ResultStage 0 (MapPartitionsRDD[1] at map at SparkPi.scala:34), which has no missing parents
20/09/08 08:52:32 INFO MemoryStore: Block broadcast_0 stored as values in memory (estimated size 3.1 KiB, free 366.3 MiB)
20/09/08 08:52:32 INFO MemoryStore: Block broadcast_0_piece0 stored as bytes in memory (estimated size 1816.0 B, free 366.3 MiB)
20/09/08 08:52:32 INFO BlockManagerInfo: Added broadcast_0_piece0 in memory on 192.168.0.100:62184 (size: 1816.0 B, free: 366.3 MiB)
20/09/08 08:52:32 INFO SparkContext: Created broadcast 0 from broadcast at DAGScheduler.scala:1200
20/09/08 08:52:32 INFO DAGScheduler: Submitting 2 missing tasks from ResultStage 0 (MapPartitionsRDD[1] at map at SparkPi.scala:34) (first 15 tasks are for partitions Vector(0, 1))
20/09/08 08:52:32 INFO TaskSchedulerImpl: Adding task set 0.0 with 2 tasks
20/09/08 08:52:34 INFO ResourceProfile: Default ResourceProfile created, executor resources: Map(cores -> name: cores, amount: 1, script: , vendor: , memory -> name: memory, amount: 1024, script: , vendor: ), task resources: Map(cpus -> name: cpus, amount: 1.0)
20/09/08 08:52:35 INFO CoarseGrainedSchedulerBackend$DriverEndpoint: Registered executor NettyRpcEndpointRef(spark-client://Executor) (192.168.0.100:62186) with ID 0
20/09/08 08:52:35 INFO BlockManagerMasterEndpoint: Registering block manager 192.168.0.100:62188 with 366.3 MiB RAM, BlockManagerId(0, 192.168.0.100, 62188, None)
20/09/08 08:52:35 INFO TaskSetManager: Starting task 0.0 in stage 0.0 (TID 0, 192.168.0.100, executor 0, partition 0, PROCESS_LOCAL, 7397 bytes)
20/09/08 08:52:35 INFO TaskSetManager: Starting task 1.0 in stage 0.0 (TID 1, 192.168.0.100, executor 0, partition 1, PROCESS_LOCAL, 7397 bytes)
20/09/08 08:52:35 INFO BlockManagerInfo: Added broadcast_0_piece0 in memory on 192.168.0.100:62188 (size: 1816.0 B, free: 366.3 MiB)
20/09/08 08:52:36 INFO TaskSetManager: Finished task 0.0 in stage 0.0 (TID 0) in 1387 ms on 192.168.0.100 (executor 0) (1/2)
20/09/08 08:52:36 INFO TaskSetManager: Finished task 1.0 in stage 0.0 (TID 1) in 1340 ms on 192.168.0.100 (executor 0) (2/2)
20/09/08 08:52:36 INFO TaskSchedulerImpl: Removed TaskSet 0.0, whose tasks have all completed, from pool
20/09/08 08:52:36 INFO DAGScheduler: ResultStage 0 (reduce at SparkPi.scala:38) finished in 4.205 s
20/09/08 08:52:36 INFO DAGScheduler: Job 0 is finished. Cancelling potential speculative or zombie tasks for this job
20/09/08 08:52:36 INFO TaskSchedulerImpl: Killing all running tasks in stage 0: Stage finished
20/09/08 08:52:36 INFO DAGScheduler: Job 0 finished: reduce at SparkPi.scala:38, took 4.375118 s
Pi is roughly 3.1478357391786957
20/09/08 08:52:36 INFO SparkUI: Stopped Spark web UI at http://192.168.0.100:4040
20/09/08 08:52:36 INFO StandaloneSchedulerBackend: Shutting down all executors
20/09/08 08:52:36 INFO CoarseGrainedSchedulerBackend$DriverEndpoint: Asking each executor to shut down
20/09/08 08:52:36 INFO MapOutputTrackerMasterEndpoint: MapOutputTrackerMasterEndpoint stopped!
20/09/08 08:52:36 INFO MemoryStore: MemoryStore cleared
20/09/08 08:52:36 INFO BlockManager: BlockManager stopped
20/09/08 08:52:36 INFO BlockManagerMaster: BlockManagerMaster stopped
20/09/08 08:52:36 INFO OutputCommitCoordinator$OutputCommitCoordinatorEndpoint: OutputCommitCoordinator stopped!
20/09/08 08:52:36 INFO SparkContext: Successfully stopped SparkContext
20/09/08 08:52:36 INFO ShutdownHookManager: Shutdown hook called
20/09/08 08:52:37 INFO ShutdownHookManager: Deleting directory /private/var/folders/dg/b9vzh3ls5d57qmmffjpvc8540000gn/T/spark-3c24377c-36fe-46d7-b948-3a6f8321451a
20/09/08 08:52:37 INFO ShutdownHookManager: Deleting directory /private/var/folders/dg/b9vzh3ls5d57qmmffjpvc8540000gn/T/spark-ea62154a-ece4-4631-adb5-cc971e3d2b1f
```

srowen commented:
Jenkins test this please

SparkQA commented:
**[Test build #128410 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/128410/testReport)** for PR 29653 at commit [`2c8c4bb`](https://github.com/apache/spark/commit/2c8c4bb6542d6d8a9b117758cec8f5416ebf023d).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/33035/
Test PASSed.

SparkQA commented:
**[Test build #128410 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/128410/testReport)** for PR 29653 at commit [`2c8c4bb`](https://github.com/apache/spark/commit/2c8c4bb6542d6d8a9b117758cec8f5416ebf023d).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/128410/
Test FAILed.

srowen commented:
Jenkins retest this please

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/33042/
Test PASSed.

SparkQA commented:
**[Test build #128419 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/128419/testReport)** for PR 29653 at commit [`2c8c4bb`](https://github.com/apache/spark/commit/2c8c4bb6542d6d8a9b117758cec8f5416ebf023d).

SparkQA commented:
**[Test build #128419 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/128419/testReport)** for PR 29653 at commit [`2c8c4bb`](https://github.com/apache/spark/commit/2c8c4bb6542d6d8a9b117758cec8f5416ebf023d).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/128419/
Test FAILed.

KevinSmile commented:
Jenkins retest this please

KevinSmile commented:
It seems that the Kafka environment of Jenkins-ci is unstable:

![image](https://user-images.githubusercontent.com/17903517/92590909-9595dc80-f2cf-11ea-83c0-313c9ccdb36e.png)

srowen commented:
Yes it's unrelated. I can try again

srowen commented:
Jenkins retest this please

SparkQA commented:
**[Test build #128451 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/128451/testReport)** for PR 29653 at commit [`2c8c4bb`](https://github.com/apache/spark/commit/2c8c4bb6542d6d8a9b117758cec8f5416ebf023d).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/33075/
Test PASSed.

SparkQA commented:
**[Test build #128451 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/128451/testReport)** for PR 29653 at commit [`2c8c4bb`](https://github.com/apache/spark/commit/2c8c4bb6542d6d8a9b117758cec8f5416ebf023d).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/128451/
Test FAILed.

KevinSmile commented:
It seems that `BlockManagerDecommissionIntegrationSuite` is unstable, as it passed in a previous ci but failed this time.

https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/128419/testReport/org.apache.spark.storage/BlockManagerDecommissionIntegrationSuite/

https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/128451/testReport/org.apache.spark.storage/BlockManagerDecommissionIntegrationSuite/

srowen commented:
Yes these must be unrelated. Don't worry either the test will get fixed and/or it will happen to pass soon. I'll keep an eye on it.

srowen commented:
Jenkins retest this please

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/33194/
Test PASSed.

SparkQA commented:
**[Test build #128569 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/128569/testReport)** for PR 29653 at commit [`2c8c4bb`](https://github.com/apache/spark/commit/2c8c4bb6542d6d8a9b117758cec8f5416ebf023d).

SparkQA commented:
**[Test build #128569 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/128569/testReport)** for PR 29653 at commit [`2c8c4bb`](https://github.com/apache/spark/commit/2c8c4bb6542d6d8a9b117758cec8f5416ebf023d).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/128569/
Test PASSed.

srowen commented:
Merged to master

KevinSmile commented:
@srowen Thanks!

koertkuipers commented:
i suspect this pullreq causes spark master to fail for me.
i run:
`$ mvn clean test -fae`

i see these failures:
```
[ERROR] Failures:                                                                                                                                             [609/9854]
[ERROR]   SparkSubmitCommandBuilderSuite.testExamplesRunnerWithMasterNoMainClass                                                                                        
Expected: (an instance of java.lang.IllegalArgumentException and exception with message a string containing "Missing example class name.")                              
     but: an instance of java.lang.IllegalArgumentException <java.lang.IllegalStateException: Failed to find examples' main app jar.> is a java.lang.IllegalStateExcept$
on                                                                                                                                                                      
Stacktrace was: java.lang.IllegalStateException: Failed to find examples' main app jar.                                                                                 
        at org.apache.spark.launcher.SparkSubmitCommandBuilder.findExamplesAppJar(SparkSubmitCommandBuilder.java:412)                                                   
        at org.apache.spark.launcher.SparkSubmitCommandBuilder.<init>(SparkSubmitCommandBuilder.java:142)                                                               
        at org.apache.spark.launcher.SparkSubmitCommandBuilderSuite.newCommandBuilder(SparkSubmitCommandBuilderSuite.java:423)                                          
        at org.apache.spark.launcher.SparkSubmitCommandBuilderSuite.buildCommand(SparkSubmitCommandBuilderSuite.java:429)                                               
        at org.apache.spark.launcher.SparkSubmitCommandBuilderSuite.testExamplesRunnerWithMasterNoMainClass(SparkSubmitCommandBuilderSuite.java:227)                    
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)                                                                                                  
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)                                                                                
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)                                                                        
        at java.lang.reflect.Method.invoke(Method.java:498)                                                                                                             
        at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)                                                                         
        at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)                                                                          
        at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)                                                                           
        at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)                                                                            
        at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)                                                                                  
        at org.junit.rules.ExpectedException$ExpectedExceptionStatement.evaluate(ExpectedException.java:239)                                                            
        at org.junit.rules.RunRules.evaluate(RunRules.java:20)                                                                                                          
        at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)                                                                                                
        at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)                                                                            
        at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)                                                                            
        at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)                                                                                                  
        at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
        at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
        at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
        at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
        at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
        at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
        at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
        at org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:365)
        at org.apache.maven.surefire.junit4.JUnit4Provider.executeWithRerun(JUnit4Provider.java:273)
        at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:238)
        at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:159)
        at org.apache.maven.surefire.booter.ForkedBooter.invokeProviderInSameClassLoader(ForkedBooter.java:384)
        at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:345)
        at org.apache.maven.surefire.booter.ForkedBooter.execute(ForkedBooter.java:126)
        at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:418)

ERROR] Errors: 
[ERROR]   SparkSubmitCommandBuilderSuite.testCliKillAndStatus:87->testCLIOpts:442->buildCommand:429->newCommandBuilder:423 » IllegalState
[ERROR]   SparkSubmitCommandBuilderSuite.testExamplesRunner:240->buildCommand:429->newCommandBuilder:423 » IllegalState
[ERROR]   SparkSubmitCommandBuilderSuite.testExamplesRunnerNoArg »  Unexpected exception...
[ERROR]   SparkSubmitCommandBuilderSuite.testExamplesRunnerNoMainClass:212->testCLIOpts:442->buildCommand:429->newCommandBuilder:423 » IllegalState
[ERROR]   SparkSubmitCommandBuilderSuite.testExamplesRunnerPrimaryResource:257->newCommandBuilder:423 » IllegalState

[ERROR] Tests run: 46, Failures: 1, Errors: 5, Skipped: 0
```

and

```
INFO] Reactor Summary for Spark Project Parent POM 3.1.0-SNAPSHOT:                                                                                                     
[INFO]                                                                                                                                                                  
[INFO] Spark Project Parent POM ........................... SUCCESS [  5.113 s]                                                                                         
[INFO] Spark Project Tags ................................. SUCCESS [ 12.398 s]                                                                                         
[INFO] Spark Project Sketch ............................... SUCCESS [ 35.952 s]                                                                                         
[INFO] Spark Project Local DB ............................. SUCCESS [  8.769 s]                                                                                         
[INFO] Spark Project Networking ........................... SUCCESS [01:01 min]                                                                                         
[INFO] Spark Project Shuffle Streaming Service ............ SUCCESS [  9.461 s]                                                                                         
[INFO] Spark Project Unsafe ............................... SUCCESS [ 19.411 s]                                                                                         
[INFO] Spark Project Launcher ............................. FAILURE [  5.735 s]                                                                                         
[INFO] Spark Project Core ................................. SKIPPED                                                                                                     
[INFO] Spark Project ML Local Library ..................... SUCCESS [ 53.520 s]                                                                                         
[INFO] Spark Project GraphX ............................... SKIPPED                                                                                                     
[INFO] Spark Project Streaming ............................ SKIPPED                                                                                                     
[INFO] Spark Project Catalyst ............................. SKIPPED                                                                                                     
[INFO] Spark Project SQL .................................. SKIPPED                                                                                                     
[INFO] Spark Project ML Library ........................... SKIPPED                                                                                                     
[INFO] Spark Project Tools ................................ SUCCESS [ 11.154 s]                                                                                         
[INFO] Spark Project Hive ................................. SKIPPED                                                                                                     
[INFO] Spark Project REPL ................................. SKIPPED                                                                                                     
[INFO] Spark Project Assembly ............................. SKIPPED                                                                                                     
[INFO] Kafka 0.10+ Token Provider for Streaming ........... SKIPPED                                                                                                     
[INFO] Spark Integration for Kafka 0.10 ................... SKIPPED                                                                                                     
[INFO] Kafka 0.10+ Source for Structured Streaming ........ SKIPPED                                                                                                     
[INFO] Spark Project Examples ............................. SKIPPED                                                                                                     
[INFO] Spark Integration for Kafka 0.10 Assembly .......... SKIPPED                                                                                                     
[INFO] Spark Avro ......................................... SKIPPED                                             
```

HyukjinKwon commented on launcher/src/main/java/org/apache/spark/launcher/SparkSubmitCommandBuilder.java:
```diff
@@ -401,6 +403,15 @@ private boolean isThriftServer(String mainClass) {
+        return exampleJar;
+      }
+    }
+    throw new IllegalStateException("Failed to find examples' main app jar.");
```
@koertkuipers I think you can fix it like:

```scala
if (isTesting) {
  SparkLauncher.NO_RESOURCE
} else {
  throw new IllegalStateException("Failed to find examples' main app jar.");
}
```

Are you interested in fixing it as a followup pr via testing as you did?

HyukjinKwon commented:
Yeah makes senses. In CI we build first so the jars will always be available. If we run the tests with compilation like you did, the jars might not be available.

KevinSmile commented:
Oh Thanks! 
My fault. Also need to delete the ut `SparkSubmitCommandBuilderSuite.testExamplesRunnerPrimaryResource()` as you cannot find the app-jar during test.

HyukjinKwon commented:
@KevinSmile, can you make a followup PR with the reproducible steps @koertkuipers provided above?

KevinSmile commented:
Yes, I'm on it. @HyukjinKwon @koertkuipers Thanks again!
